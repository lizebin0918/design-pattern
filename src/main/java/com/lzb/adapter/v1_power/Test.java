package com.lzb.adapter.v1_power;
/**
 * <br/>
 * Created on : 2021-05-15 23:11
 * @author lizebin
 */
public class Test {

    public static void main(String[] args) {
        PowerDC5 powerDC5 = new PowerAdapter(new Power220V());
        powerDC5.output();
    }

}
