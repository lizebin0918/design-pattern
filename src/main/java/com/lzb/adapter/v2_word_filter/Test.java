package com.lzb.adapter.v2_word_filter;

import com.lzb.adapter.v2_word_filter.adapter.*;

/**
 * <br/>
 * Created on : 2021-05-16 10:43
 * @author lizebin
 */
public class Test {

    public static void main(String[] args) {
        FilterManagement filterManagement = new FilterManagement();
        filterManagement.add(new ASensitiveWordsFilterAdaptor());
        filterManagement.add(new BSensitiveWordsFilterAdaptor());
        filterManagement.add(new CSensitiveWordsFilterAdaptor());
        System.out.println(filterManagement.filter("梁小姐~~~哈哈哈哈"));
    }

}
