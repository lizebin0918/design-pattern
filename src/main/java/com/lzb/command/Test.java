package com.lzb.command;

import java.util.function.Function;

/**
 * 命令模式<br/>
 * Created on : 2021-05-16 22:08
 * @author lizebin
 */
public class Test {

    private static class Request {
        private Function<Integer, Integer> func;
        private int number;

        public Request(Function<Integer, Integer> func, int number) {
            this.func = func;
            this.number = number;
        }

        void handle() {
            System.out.println(func.apply(number));
        }
    }

    public static void main(String[] args) {
        new Request(n -> n + 10, 90).handle();
    }

}
