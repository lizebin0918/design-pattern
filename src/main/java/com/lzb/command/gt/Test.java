package com.lzb.command.gt;

import lombok.Data;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * 命令模式将请求（命令）封装成一个对象，这样可以使用不同的请求参数化其他对象，并且能够支持请求（命令）的排队执行：动作 + 数据
 *
 * 当我们把函数封装成对象之后，对象可以存储下来、方便控制。主要应用场景：用来控制命令的执行，这里的执行可以是同步、异步、延迟、排队。
 * 消费者无需关注命令的执行逻辑，只关注执行命令的逻辑
 *
 * 命令模式和策略模式的区别：
 * 策略模式：不同的策略具有相同的目的、不同的实现、互相之间可以替换。比如冒泡排序、选择排序都是为了实现排序，它们的最终目标都是实现排序而已
 * 命令模式：不同的命令具有不同的目的，对应不同的处理逻辑，并且相互之间不可替换
 *
 * 想当年对接多个ERP接口，客户端定义固定接口，采用策略+模板方式组织流程，每次新来一套接口，直接加实现即可。还可以有另一种方式，采用命令模式，这样还能解耦，
 * 生产者只要把命令（动作+数据）封装好放到队列，消费者消费即可，但是这种方式有可能导致一个指令的消费阻塞导致后面的排队指令消费不了
 *
 * Created on : 2021-10-15 13:31
 *
 * @author lizebin
 */
public class Test {

    private static interface Command {
        void execute();
    }

    @Data
    private static class GotDiamondCommand implements Command {

        private String data;
        public GotDiamondCommand(String data) {
            this.data = data;
        }

        @Override
        public void execute() {
            System.out.println("获取钻石执行");
        }
    }

    @Data
    private static class Request {
        private Event event;
        private String data;
    }

    /**
     * 事件枚举
     */
    private enum Event {
        GOT_DIAMOND;
    }

    private static final int MAX_HANDLER_REQ_COUNT_PER_LOOP = 100;

    private Queue<Command> queue = new LinkedList<>();

    public void mainloop() {
        while (true) {
            List<Request> requests = new ArrayList<>();
            requests.forEach(request -> {
                Event event = request.getEvent();
                Command command = null;
                if (event.equals(Event.GOT_DIAMOND)) {
                    command = new GotDiamondCommand(request.getData());
                } // else if ()
                queue.add(command);
            });

            int handledCount = 0;
            while (handledCount < MAX_HANDLER_REQ_COUNT_PER_LOOP) {
                if (queue.isEmpty()) {
                    break;
                }
                Command command = queue.poll();
                command.execute();
                ++handledCount;
            }
        }
    }

}
