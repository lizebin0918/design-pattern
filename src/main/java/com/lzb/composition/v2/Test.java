package com.lzb.composition.v2;
/**
 * <br/>
 * Created on : 2021-05-16 17:47
 * @author lizebin
 */
public class Test {

    public static void main(String[] args) {
        MyDirectory fileDir = new MyDirectory("/Users/lizebin/Desktop/temp");
        MyDirectory nodeX = new MyDirectory("/Users/lizebin/Desktop/temp/x");
        MyDirectory nodeY = new MyDirectory("/Users/lizebin/Desktop/temp/y");
        fileDir.addSubNode(nodeX);
        fileDir.addSubNode(nodeY);


        nodeX.addSubNode(new MyFile("/Users/lizebin/Desktop/temp/x/x.txt"));
        nodeX.addSubNode(new MyFile("/Users/lizebin/Desktop/temp/x/x1.txt"));
        nodeY.addSubNode(new MyFile("/Users/lizebin/Desktop/temp/y/y.txt"));
        nodeY.addSubNode(new MyFile("/Users/lizebin/Desktop/temp/y/y1.txt"));

        System.out.println(fileDir.countNumOfFiles());
        System.out.println(fileDir.countSizeOfFiles());
    }

}
