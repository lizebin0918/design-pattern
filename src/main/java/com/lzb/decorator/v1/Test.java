package com.lzb.decorator.v1;
/**
 * <br/>
 * Created on : 2021-05-15 20:32
 * @author lizebin
 */
public class Test {

    public static void main(String[] args) {
        PanCake panCake = new PanCake();
        System.out.println(panCake.getMsg() + ",总价:" + panCake.getPrice());

        PanCake panCakeWithEgg = new PanCakeWithEgg();
        System.out.println(panCakeWithEgg.getMsg() + ",总价:" + panCakeWithEgg.getPrice());

        PanCake panCakeWithEggAndSausage = new PanCakeWithEggAndSausage();
        System.out.println(panCakeWithEggAndSausage.getMsg() + ",总价:" + panCakeWithEggAndSausage.getPrice());
    }

}
