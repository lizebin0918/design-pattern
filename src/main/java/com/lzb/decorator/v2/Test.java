package com.lzb.decorator.v2;

import lombok.Data;

/**
 * <br/>
 * Created on : 2021-05-15 20:57
 * @author lizebin
 */
public class Test {

    public static void main(String[] args) {
        PanCake panCake = new BasePanCake();
        panCake = new EggDecorator(panCake);
        panCake = new EggDecorator(panCake);
        panCake = new SausageDecorator(panCake);
        System.out.println(panCake.getMsg());
        System.out.println("总价:" + panCake.getPrice());

        Person lizebin = new Lizebin("lizebin");
        System.out.println(lizebin.getName());

        // 变态的多态：https://wiki.jikexueyuan.com/project/java-enhancement/java-three.html
        A a1 = new A();
        A a2 = new B();
        B b = new B();
        C c = new C();
        D d = new D();

        System.out.println("1--" + a1.show(b));
        System.out.println("2--" + a1.show(c));
        System.out.println("3--" + a1.show(d));
        System.out.println("4--" + a2.show(b));
        System.out.println("5--" + a2.show(c));
        System.out.println("6--" + a2.show(d));
        System.out.println("7--" + b.show(b));
        System.out.println("8--" + b.show(c));
        System.out.println("9--" + b.show(d));
    }

    @Data
    public static abstract class Person {

        private String name;

        public Person(String name) {
            this.name = name;
        }

    }

    public static class Lizebin extends Person {

        public Lizebin(String name) {
            super(name);
        }

        @Override
        public String getName() {
            return getMyName() + super.getName();
        }

        public String getMyName() {
            return "myName:";
        }

    }

    public static class A {

        public String show(D obj) {
            return ("A and D");
        }

        public String show(A obj) {
            return ("A and A");
        }
    }

    public static class B extends A {
        public String show(B obj){
            return ("B and B");
        }

        @Override
        public String show(A obj){
            return ("B and A");
        }
    }

    public static class C extends B{

    }

    public static class D extends B{

    }

}
