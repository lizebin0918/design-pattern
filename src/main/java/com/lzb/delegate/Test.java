package com.lzb.delegate;
/**
 * <br/>
 * Created on : 2021-05-15 21:38
 * @author lizebin
 */
public class Test {

    public static void main(String[] args) {
        Boss boss = new Boss();
        Manager manager = new Manager();
        boss.command("A_Module", manager);
    }

}
