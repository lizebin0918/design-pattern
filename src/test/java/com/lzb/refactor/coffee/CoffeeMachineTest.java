package com.lzb.refactor.coffee;

import com.lzb.SystemOutUtils;
import org.approvaltests.ApprovalUtilities;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import java.io.ByteArrayOutputStream;

class CoffeeMachineTest {

    @Test
    void test_one() {
        ByteArrayOutputStream output = new ApprovalUtilities().writeSystemOutToStringBuffer();
        CoffeeMachineMock.run("1");
        CoffeeMachineMock.run("2");
        CoffeeMachineMock.run("3");
        CoffeeMachineMock.run("4");
        CoffeeMachineMock.run("5");
        CoffeeMachineMock.run("6");
        Approvals.verify(SystemOutUtils.convert(output));
    }

    @Test
    void test_two() {
        ByteArrayOutputStream output = new ApprovalUtilities().writeSystemOutToStringBuffer();
        CoffeeMachineMock.run("1");
        CoffeeMachineMock.run("1");
        CoffeeMachineMock.run("1");
        Approvals.verify(SystemOutUtils.convert(output));
    }

    @Test
    void test_refresh() {
        ByteArrayOutputStream output = new ApprovalUtilities().writeSystemOutToStringBuffer();
        CoffeeMachineMock.run("r");
        Approvals.verify(SystemOutUtils.convert(output));
    }

}