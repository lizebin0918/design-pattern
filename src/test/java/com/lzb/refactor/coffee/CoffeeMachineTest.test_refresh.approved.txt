Inventory:
Cocoa,10
Cocoa,10
Cocoa,10
Cocoa,10
Cocoa,10
Cocoa,10
Cocoa,10
Cocoa,10
Cocoa,10
Cocoa,10
Coffee,10
Coffee,10
Coffee,10
Coffee,10
Coffee,10
Coffee,10
Coffee,10
Coffee,10
Coffee,10
Coffee,10
Cream,10
Cream,10
Cream,10
Cream,10
Cream,10
Cream,10
Cream,10
Cream,10
Cream,10
Cream,10
Decaf Coffee,10
Decaf Coffee,10
Decaf Coffee,10
Decaf Coffee,10
Decaf Coffee,10
Decaf Coffee,10
Decaf Coffee,10
Decaf Coffee,10
Decaf Coffee,10
Decaf Coffee,10
Espresso,1
Espresso,4
Espresso,7
Espresso,10
Espresso,10
Espresso,10
Espresso,10
Espresso,10
Espresso,10
Espresso,10
Foamed Milk,10
Foamed Milk,10
Foamed Milk,10
Foamed Milk,10
Foamed Milk,10
Foamed Milk,10
Foamed Milk,10
Foamed Milk,10
Foamed Milk,10
Foamed Milk,10
Steamed Milk,10
Steamed Milk,10
Steamed Milk,10
Steamed Milk,10
Steamed Milk,10
Steamed Milk,10
Steamed Milk,10
Steamed Milk,10
Steamed Milk,10
Steamed Milk,10
Sugar,10
Sugar,10
Sugar,10
Sugar,10
Sugar,10
Sugar,10
Sugar,10
Sugar,10
Sugar,10
Sugar,10
Whipped Cream,10
Whipped Cream,10
Whipped Cream,10
Whipped Cream,10
Whipped Cream,10
Whipped Cream,10
Whipped Cream,10
Whipped Cream,10
Whipped Cream,10
Whipped Cream,10

Menu:
1,Caffe Americano,$33.00,false
2,Caffe Americano,$33.00,false
3,Caffe Americano,$33.00,false
4,Caffe Americano,$33.00,false
5,Caffe Americano,$33.00,false
6,Caffe Americano,$33.00,false
7,Caffe Americano,$33.00,false
8,Caffe Americano,$33.00,false
9,Caffe Americano,$33.00,false
10,Caffe Americano,$33.00,false
11,Caffe Latte,$25.50,false
12,Caffe Latte,$25.50,false
13,Caffe Latte,$25.50,false
14,Caffe Latte,$25.50,false
15,Caffe Latte,$25.50,false
16,Caffe Latte,$25.50,false
17,Caffe Latte,$25.50,false
18,Caffe Latte,$25.50,false
19,Caffe Latte,$25.50,false
20,Caffe Latte,$25.50,false
21,Caffe Mocha,$33.50,true
22,Caffe Mocha,$33.50,true
23,Caffe Mocha,$33.50,true
24,Caffe Mocha,$33.50,true
25,Caffe Mocha,$33.50,true
26,Caffe Mocha,$33.50,true
27,Caffe Mocha,$33.50,true
28,Caffe Mocha,$33.50,true
29,Caffe Mocha,$33.50,true
30,Caffe Mocha,$33.50,true
31,Cappuccino,$29.00,false
32,Cappuccino,$29.00,false
33,Cappuccino,$29.00,false
34,Cappuccino,$29.00,false
35,Cappuccino,$29.00,false
36,Cappuccino,$29.00,false
37,Cappuccino,$29.00,false
38,Cappuccino,$29.00,false
39,Cappuccino,$29.00,false
40,Cappuccino,$29.00,false
41,Coffee,$27.50,true
42,Coffee,$27.50,true
43,Coffee,$27.50,true
44,Coffee,$27.50,true
45,Coffee,$27.50,true
46,Coffee,$27.50,true
47,Coffee,$27.50,true
48,Coffee,$27.50,true
49,Coffee,$27.50,true
50,Coffee,$27.50,true
51,Decaf Coffee,$27.50,true
52,Decaf Coffee,$27.50,true
53,Decaf Coffee,$27.50,true
54,Decaf Coffee,$27.50,true
55,Decaf Coffee,$27.50,true
56,Decaf Coffee,$27.50,true
57,Decaf Coffee,$27.50,true
58,Decaf Coffee,$27.50,true
59,Decaf Coffee,$27.50,true
60,Decaf Coffee,$27.50,true

Your selection: Inventory:
Cocoa,10
Cocoa,10
Cocoa,10
Cocoa,10
Cocoa,10
Cocoa,10
Cocoa,10
Cocoa,10
Cocoa,10
Cocoa,10
Coffee,10
Coffee,10
Coffee,10
Coffee,10
Coffee,10
Coffee,10
Coffee,10
Coffee,10
Coffee,10
Coffee,10
Cream,10
Cream,10
Cream,10
Cream,10
Cream,10
Cream,10
Cream,10
Cream,10
Cream,10
Cream,10
Decaf Coffee,10
Decaf Coffee,10
Decaf Coffee,10
Decaf Coffee,10
Decaf Coffee,10
Decaf Coffee,10
Decaf Coffee,10
Decaf Coffee,10
Decaf Coffee,10
Decaf Coffee,10
Espresso,10
Espresso,10
Espresso,10
Espresso,10
Espresso,10
Espresso,10
Espresso,10
Espresso,10
Espresso,10
Espresso,10
Foamed Milk,10
Foamed Milk,10
Foamed Milk,10
Foamed Milk,10
Foamed Milk,10
Foamed Milk,10
Foamed Milk,10
Foamed Milk,10
Foamed Milk,10
Foamed Milk,10
Steamed Milk,10
Steamed Milk,10
Steamed Milk,10
Steamed Milk,10
Steamed Milk,10
Steamed Milk,10
Steamed Milk,10
Steamed Milk,10
Steamed Milk,10
Steamed Milk,10
Sugar,10
Sugar,10
Sugar,10
Sugar,10
Sugar,10
Sugar,10
Sugar,10
Sugar,10
Sugar,10
Sugar,10
Whipped Cream,10
Whipped Cream,10
Whipped Cream,10
Whipped Cream,10
Whipped Cream,10
Whipped Cream,10
Whipped Cream,10
Whipped Cream,10
Whipped Cream,10
Whipped Cream,10

Menu:
1,Caffe Americano,$33.00,true
2,Caffe Americano,$33.00,true
3,Caffe Americano,$33.00,true
4,Caffe Americano,$33.00,true
5,Caffe Americano,$33.00,true
6,Caffe Americano,$33.00,true
7,Caffe Americano,$33.00,true
8,Caffe Americano,$33.00,true
9,Caffe Americano,$33.00,true
10,Caffe Americano,$33.00,true
11,Caffe Latte,$25.50,true
12,Caffe Latte,$25.50,true
13,Caffe Latte,$25.50,true
14,Caffe Latte,$25.50,true
15,Caffe Latte,$25.50,true
16,Caffe Latte,$25.50,true
17,Caffe Latte,$25.50,true
18,Caffe Latte,$25.50,true
19,Caffe Latte,$25.50,true
20,Caffe Latte,$25.50,true
21,Caffe Mocha,$33.50,true
22,Caffe Mocha,$33.50,true
23,Caffe Mocha,$33.50,true
24,Caffe Mocha,$33.50,true
25,Caffe Mocha,$33.50,true
26,Caffe Mocha,$33.50,true
27,Caffe Mocha,$33.50,true
28,Caffe Mocha,$33.50,true
29,Caffe Mocha,$33.50,true
30,Caffe Mocha,$33.50,true
31,Cappuccino,$29.00,true
32,Cappuccino,$29.00,true
33,Cappuccino,$29.00,true
34,Cappuccino,$29.00,true
35,Cappuccino,$29.00,true
36,Cappuccino,$29.00,true
37,Cappuccino,$29.00,true
38,Cappuccino,$29.00,true
39,Cappuccino,$29.00,true
40,Cappuccino,$29.00,true
41,Coffee,$27.50,true
42,Coffee,$27.50,true
43,Coffee,$27.50,true
44,Coffee,$27.50,true
45,Coffee,$27.50,true
46,Coffee,$27.50,true
47,Coffee,$27.50,true
48,Coffee,$27.50,true
49,Coffee,$27.50,true
50,Coffee,$27.50,true
51,Decaf Coffee,$27.50,true
52,Decaf Coffee,$27.50,true
53,Decaf Coffee,$27.50,true
54,Decaf Coffee,$27.50,true
55,Decaf Coffee,$27.50,true
56,Decaf Coffee,$27.50,true
57,Decaf Coffee,$27.50,true
58,Decaf Coffee,$27.50,true
59,Decaf Coffee,$27.50,true
60,Decaf Coffee,$27.50,true

Your selection: 
